File: android/app/src/main/java/com/vetdose/app/data/settings/SettingsRepository.kt
package com.vetdose.app.data.settings

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

private val Context.dataStore by preferencesDataStore(name = "vetdose_settings")

/**
 * Persists the two pieces of state the sync flow needs across process death:
 * where the server is, and when we last synced successfully (the `since`
 * cursor for the next incremental `/sync` call).
 */
@Singleton
class SettingsRepository @Inject constructor(@ApplicationContext private val context: Context) {

    private object Keys {
        val SERVER_BASE_URL = stringPreferencesKey("server_base_url")
        val LAST_SYNC_TIME = stringPreferencesKey("last_sync_time")
    }

    val serverBaseUrlFlow: Flow<String> =
        context.dataStore.data.map { it[Keys.SERVER_BASE_URL] ?: DEFAULT_SERVER_BASE_URL }

    val lastSyncTimeFlow: Flow<String?> =
        context.dataStore.data.map { it[Keys.LAST_SYNC_TIME] }

    suspend fun currentServerBaseUrl(): String = serverBaseUrlFlow.first()

    suspend fun setServerBaseUrl(url: String) {
        context.dataStore.edit { it[Keys.SERVER_BASE_URL] = url }
    }

    suspend fun setLastSyncTime(isoTimestamp: String) {
        context.dataStore.edit { it[Keys.LAST_SYNC_TIME] = isoTimestamp }
    }

    companion object {
        // 10.0.2.2 is the Android emulator's alias for the host machine's
        // localhost, where `docker compose up` runs the backend in dev.
        const val DEFAULT_SERVER_BASE_URL = "http://10.0.2.2:8000/"
    }
}

File: web/src/data/settingsRepository.ts
/** Small persisted app settings — the web equivalent of Android's
 * DataStore-backed `SettingsRepository`: the backend base URL and the `since`
 * cursor for the next incremental `/sync`. Backed by `db.settings`
 * (`data/local/db.ts`) rather than `localStorage` so it stays inside the same
 * transactional store as the synced data. */

import { db } from "./local/db"

const SERVER_BASE_URL_KEY = "serverBaseUrl"
const LAST_SYNC_TIME_KEY = "lastSyncTime"

/** Vite's dev server runs on :5173 (see `backend/app/config.py`'s default
 * CORS origin); the backend itself defaults to :8000. */
export const DEFAULT_SERVER_BASE_URL = "http://localhost:8000/"

export async function getServerBaseUrl(): Promise<string> {
  const row = await db.settings.get(SERVER_BASE_URL_KEY)
  return row?.value ?? DEFAULT_SERVER_BASE_URL
}

export async function setServerBaseUrl(url: string): Promise<void> {
  await db.settings.put({ key: SERVER_BASE_URL_KEY, value: url })
}

/** ISO-8601 timestamp of the last successful sync, or `null` before the
 * first one — passed as `?since=` to make the next `/sync` incremental. */
export async function getLastSyncTime(): Promise<string | null> {
  const row = await db.settings.get(LAST_SYNC_TIME_KEY)
  return row?.value ?? null
}

export async function setLastSyncTime(isoTimestamp: string): Promise<void> {
  await db.settings.put({ key: LAST_SYNC_TIME_KEY, value: isoTimestamp })
}
